docs/data/data-grid/pagination/ServerPaginationGridEstimated.tsx

```diff
@@ -0,0 +1,66 @@
+import * as React from 'react';
+import Button from '@mui/material/Button';
+import { DataGrid, useGridApiRef } from '@mui/x-data-grid';
+import type { GridPaginationMeta, GridPaginationModel } from '@mui/x-data-grid';
+import { createFakeServer } from '@mui/x-data-grid-generator';
+
+const SERVER_OPTIONS = {
+  useCursorPagination: false,
+};
+
+const { useQuery, ...data } = createFakeServer({ rowLength: 1000 }, SERVER_OPTIONS);
+
+export default function ServerPaginationGridEstimated() {
+  const apiRef = useGridApiRef();
+  const [paginationModel, setPaginationModel] = React.useState({
+    page: 0,
+    pageSize: 50,
+  });
+
+  const {
+    isLoading,
+    rows,
+    pageInfo: { hasNextPage },
+  } = useQuery(paginationModel);
+
+  const [paginationMeta, setPaginationMeta] = React.useState<GridPaginationMeta>({});
+
+  React.useEffect(() => {
+    if (hasNextPage !== undefined) {
+      setPaginationMeta((prev) => {
+        if (prev.hasNextPage !== hasNextPage) {
+          return { ...prev, hasNextPage };
+        }
+        return prev;
+      });
+    }
+  }, [hasNextPage]);
```

Comment:
```suggestion
  const paginationMetaRef = React.useRef<GridPaginationMeta>({});
  // Memoize to avoid flickering when the `hasNextPage` is `undefined` during refetch
  const paginationMeta = React.useMemo(() => {
    if (
      hasNextPage !== undefined &&
      paginationMetaRef.current?.hasNextPage !== hasNextPage
    ) {
      paginationMetaRef.current = { hasNextPage };
    }
    return paginationMetaRef.current;
  }, [hasNextPage]);
```